jupyterlab/handlers/yjs_echo_ws.py

```diff
@@ -101,30 +114,94 @@ def on_message(self, message):
         elif message[0] == ServerMessageType.RENAME_SESSION:
             # We move the room to a different entry and also change the room_id property of each connected client
             new_room_id = message[1:].decode("utf-8")
-            for client_id, (loop, hook_send_message, client) in room.clients.items() :
+            for client_id, (loop, hook_send_message, client) in room.clients.items():
                 client.room_id = new_room_id
-            cls.rooms.pop(room_id)
-            cls.rooms[new_room_id] = room
+            ROOMS.pop(room_id)
+            ROOMS[new_room_id] = room
             # print("renamed room to " + new_room_id + ". Old room name was " + room_id)
         elif room:
-            for client_id, (loop, hook_send_message, client) in room.clients.items() :
-                if self.id != client_id :
+            if message[0] == 0:  # sync message
+                read_sync_message(self, room.ydoc.ydoc, message[1:])
+            for client_id, (loop, hook_send_message, client) in room.clients.items():
+                if self.id != client_id:
                     loop.add_callback(hook_send_message, message)
 
     def on_close(self):
         # print("[YJSEchoWS]: close")
-        cls = self.__class__
-        room = cls.rooms.get(self.room_id)
+        room = ROOMS.get(self.room_id)
         room.clients.pop(self.id)
-        if len(room.clients) == 0 :
-            cls.rooms.pop(self.room_id)
+        if not room.clients:
+            ROOMS.pop(self.room_id)
             # print("[YJSEchoWS]: close room " + self.room_id)
 
         return True
 
     def check_origin(self, origin):
-        #print("[YJSEchoWS]: check origin")
+        # print("[YJSEchoWS]: check origin")
         return True
 
     def hook_send_message(self, msg):
         self.write_message(msg, binary=True)
+
+message_yjs_sync_step1 = 0
+message_yjs_sync_step2 = 1
+message_yjs_update = 2
+
+def read_sync_step1(handler, doc, encoded_state_vector):
+    message = Y.encode_state_as_update(doc, encoded_state_vector)
+    message = bytes([message_yjs_sync_step2] + get_len(message) + message)
```

Comment:
This is not correct. The length of the message must be encoded using variable-length encoding. I sent you some sample code earlier.

Reply:
Could you be more specific? I believe the length of the message is encoded using variable-length encoding.

Reply:
Thanks, I will rename this function.